compiler/generator_rust.c

```diff
@@ -939,6 +945,7 @@ static void generate_dollar(struct generator * g, struct node * p) {
               "~Menv.limit = env.current.len();~N", p);
     generate(g, p->left);
     if (!g->unreachable) {
+        g->V[0] = p->name;
         /* Update string variable. */
         w(g, "~M~V0 = env.current.clone().into_owned();~N");
```

Comment:
Oh, it's a bit more complex than I realised.

So in C, the `z->p` is set to `z->S[0]` so in Snowball the current string now aliases `s` (as I said), **but** if the current string is modified such that it needs to be reallocated (i.e. it gets longer than the length of the space allocated) then `z->p` will get updated to point to the reallocated block, but `z->S[0]` will not be changed, and so the aliasing is broken.  And in fact you can't safely access the string variable `s` then (until after the `$` has ended), as `z->S[0]` will point to a block that's been released.

So actually the generated C code doesn't reliably have the semantics I suggested, so the Rust backend's approach seems OK.  This restriction on not being allowed to look at the currently "active" string ought to at least be documented, and perhaps the C backend should `NULL` out the variable to help avoid accidental uses.  We could also warn about situations where this could happen at compile time fairly easily I think (basically "does any code reachable from inside this `$` touch the variable that the `$` is over.  There are already some similar checks - for example, if any code reachable from _X_ could modify the cursor.)

We could fix things so the variable either always aliases (e.g. via a double pointer indirection, or the "current active string variable" idea) or so it always keeps its original value (e.g. copy the value, or copy on write) but I'm not sure either is worth the overhead it would add to the generated code (or the effort to implement).  This is a corner case of a feature that's not used in any of the current stemmers after all.